pkg/xds/experimental/client/client.go

```diff
@@ -0,0 +1,621 @@
+	AddResourceWatcher(typeUrl string, cb WatcherCallback) uint64
+	RemoveResourceWatcher(id uint64)
+}
+
+var _ Interface = (*Client)(nil)
+
+type observeRequest struct {
+	typeUrl       string
+	resourceNames []string
+}
+
+type Client struct {
+	// node is used to identify a client to Traffic Director server.
+	// It is part of every Request sent on a stream.
+	node *corepb.Node
+	log  *slog.Logger
+	opts Options
+
+	observeQueue       chan *observeRequest
+	responseQueue      chan *discoverypb.DiscoveryResponse
+	deltaResponseQueue chan *discoverypb.DeltaDiscoveryResponse
+
+	// cache stores versioned resources.
+	cache *xds.Cache
+	// watchers manages callbacks with notification when cache state changes.
+	watchers *watchers
+}
+
+// NewClient creates a new xDS client with default options and logger.
+func NewClient(node *corepb.Node) *Client {
+	return NewClientWithOptions(slog.Default(), node, Defaults)
+}
+
+func NewClientWithOptions(log *slog.Logger, node *corepb.Node, opts *Options) *Client {
+	cache := xds.NewCache()
+
+	c := &Client{
+		node:               proto.Clone(node).(*corepb.Node),
+		log:                log,
+		opts:               *opts,
+		observeQueue:       make(chan *observeRequest, 1),
+		responseQueue:      make(chan *discoverypb.DiscoveryResponse, 1),
+		deltaResponseQueue: make(chan *discoverypb.DeltaDiscoveryResponse, 1),
+		cache:              cache,
+		watchers:           newWatchers(log.With(logFeatureKey, "watchers"), cache),
+	}
+	return c
+}
+
+// Run will start an AggregatedDiscoverService stream and process requests
+// and responses until provided Context ctx is done or non-retriable error occurs.
+func (c *Client) Run(ctx context.Context, conn grpc.ClientConnInterface) error {
+	backoff := backoff.Exponential{
+		Min:        c.opts.MinBackoff,
+		Max:        c.opts.MaxBackoff,
+		ResetAfter: c.opts.BackoffReset,
+		Jitter:     true,
+		Name:       "xds-client-conn",
+	}
+
+	for {
+		ctxConn, cancel := context.WithCancel(ctx)
+		client := discoverypb.NewAggregatedDiscoveryServiceClient(conn)
+
+		var stream discoverypb.AggregatedDiscoveryService_StreamAggregatedResourcesClient
+		var delta discoverypb.AggregatedDiscoveryService_DeltaAggregatedResourcesClient
+		var err error
+		if c.opts.UseSOTW {
+			stream, err = client.StreamAggregatedResources(ctxConn, grpc.WaitForReady(true))
+			if err != nil {
+				err = fmt.Errorf("start stream: %w", err)
+				goto retry
+			}
+		} else {
+			delta, err = client.DeltaAggregatedResources(ctxConn, grpc.WaitForReady(true))
+			if err != nil {
+				err = fmt.Errorf("start delta: %w", err)
+				goto retry
+			}
+		}
+		err = c.run(ctxConn, stream, delta)
+
+	retry:
+		cancel()
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		if !c.opts.RetryConnection {
+			return err
+		}
+		c.log.Error("Retrying connection", "error", err)
+		err = backoff.Wait(ctx)
+		if err != nil {
+			// Possible only when Context ctx is done.
```

Comment:
This comment concerns me -- it relies on knowledge of the implementation of Wait().

You should probably check that Context is actually done rather than assume it...